### ISSUE TITLE
GHCJS support?

### ISSUE THREAD

**Issue by themoritz** [commercialhaskell/intero#64]
Is intero intended to work with ghcjs projects? If so, what needs to be done to make it work? Currently I'm getting `Could not find package ghc in known packages` when running `stack build intero` in my ghcjs project.

**Comment by mgsloan** [comment]
We'd need to add some sort of flag to compile a ghcjsi version of intero.  It would add a dep on ghcjs and probably #define a CPP flag that would essentially apply luite's GHCJSi patches - https://github.com/ghcjs/ghcjs/commits/master/src/Compiler/InteractiveUI.hs .

**Comment by themoritz** [comment]
Okay thanks, I'll leave this open as a feature request. :)

**Comment by qrilka** [comment]
While this is not implemented is there any good way to disable intero autoload for ghcjs packages?
Currently I do it in not a very elegant way - open file, see intero error, switch back to the file and enter `intero-mode` to disable intero.

**Comment by chrisdone** [comment]
You should just be able to ignore the error, it shouldn't bother you further. You don't have to disable intero-mode.

**Comment by chrisdone** [comment]
But don't kill the error buffer, just ignore it.

**Comment by qrilka** [comment]
Oh, thanks @chrisdone,  probably that disabling action was meaningless, thanks, but I can't completely ignore the error as that error buffer opens up and I need to switch back to the buffer with the file I was intending to open originally.
It's not a huge problem but I was wondering if there is some way around this extra step.

**Comment by chrisdone** [comment]
I thought of a workaround, we could change the hook you use for haskell-mode to something like `'intero-mode-maybe`, which only enables `intero-mode` if there isn't a local variable disabling it.

**Comment by qrilka** [comment]
Sounds interesting

**Comment by 3noch** [comment]
I'm having a lot problems using Intero on a GHC + GHCJS project. Whenever I open a file on the GHCJS side intero repeatedly opens complaining it can't install. Even switching back to the buffer is not enough. It seems after a few tries it finally gives up. But this is a big enough issue that I mostly can't use intero on this project.

**Comment by mgsloan** [comment]
@3noch I've had issues like that too in the past, though I haven't been doing much GHCJS lately.  See https://github.com/commercialhaskell/intero/issues/88

I've also found things far less frustrating with `(setq flycheck-check-syntax-automatically '(mode-enabled save))`.  Instead of trying to compile every change, just try on save.

**Comment by qrilka** [comment]
After updating `intero` couple of days ago now I get
```
Warning (emacs): Couldn't get the Stack project root.

This can be caused by a syntax error in your stack.yaml file. Check that out.

Otherwise, please report this as a bug!
```
Appearing multiple in `*Warnings*` buffer when I e.g. save buffer from GHCJS project.
Is it some expected behavior of `intero`?

**Comment by crocket** [comment]
Currently, interactive-haskell-mode supports ghcjs repl if you have a recent version of stack.

**Comment by themoritz** [comment]
Yes, that's what I'm doing too at the moment. It's not working as smoothly as with ghc, but I think it's the best IDE option currently out there for working on GHCJS projects.

**Comment by vertexcite** [comment]
http://leksah.org/ is a dedicated Haskell IDE worth checking out for good GHCJS support.

**Comment by louispan** [comment]
@themoritz, I have been able to use intero with ghcjs project by adding [ghcjs-base-stub](https://hackage.haskell.org/package/ghcjs-base-stub) to .cabal files.

**Comment by chrisdone** [comment]
@louispan That's awesome.

**Comment by louispan** [comment]
One problem with [ghcjs-base-stub](https://hackage.haskell.org/package/ghcjs-base-stub) is that I haven't stubbed all of ghcjs-base. I'm happy to accept pull requests to add any missing modules though.

**Comment by chrisdone** [comment]
I've started using it on a project, it's wonderful! Simple solution so that we can get our regular Intero goodies.

**Comment by mgsloan** [comment]
Cool!  I'm curious how that works with ghcjs's special FFI syntax?  I'm guessing all of that needs to be shimmed out as `undefined` or `error` via CPP?

**Comment by chrisdone** [comment]
I had only a few in my Duet codebase, so I just CPP `#ifdef`'d them out. Combined with `-fdefer-type-errors`, the complaints that the FFI thing is now out of scope is just a warning that I see if I look at that module. So I'm very happy!
